tree-walk.go

```diff
@@ -37,43 +37,108 @@ type treeWalkResult struct {
 	end   bool
 }
 
+// posix.ListDir returns entries with trailing "/" for directories. At the object layer
+// we need to remove this trailing "/" for objects and retain "/" for prefixes before
+// sorting because the trailing "/" can affect the sorting results for certain cases.
+// Ex. lets say entries = ["a-b/", "a/"] and both are objects.
+//     sorting with out trailing "/" = ["a", "a-b"]
+//     sorting with trailing "/"     = ["a-b/", "a/"]
+// Hence if entries[] does not have a case like the above example then isLeaf() check
+// can be delayed till the entry is pushed into the treeWalkResult channel.
+// delayIsLeafCheck() returns true if isLeaf can be delayed or false if
+// isLeaf should be done in listDir()
+func delayIsLeafCheck(entries []string) bool {
+	for i, entry := range entries {
+		if i == len(entries)-1 {
+			break
+		}
+		// For each rune in the entry string, if any rune is less than '/'
```

Comment:
ah, I thought indexing into a string fetches a rune and not byte, but you are right:
```
krishna@escape:/tmp$ cat a.go 
package main

import "fmt"

func main() {
	a := "⌘"
	for i := range a {
		fmt.Println(a[i])
	}
	fmt.Println("====")
	b := []byte(a)
	for i := range b {
		fmt.Println(b[i])
	}
}
krishna@escape:/tmp$ go run a.go
226
====
226
140
152
krishna@escape:/tmp$ 
```